src/main/java/tfb/status/service/ShareResultsUploader.java

```diff
@@ -0,0 +1,305 @@
+      }
+
+      String resultsUrl =
+          urlsConfig.tfbStatus
+              + "/share-results/"
+              + URLEncoder.encode(fileName, UTF_8);
+      String visualizeResultsUrl =
+          urlsConfig.teWeb
+              + "/benchmarks/#section=test&shareid="
+              + URLEncoder.encode(shareId, UTF_8);
+
+      return new ShareResultsJsonView(
+          /* fileName= */ fileName,
+          /* resultsUrl= */ resultsUrl,
+          /* visualizeResultsUrl= */ visualizeResultsUrl);
+    } finally {
+      Files.delete(tempFile);
+    }
+  }
+
+  /**
+   * Return null if the JSON file successfully de-serializes to {@link Results}
+   * and has non-empty {@link Results#testMetadata}. Otherwise returns a relevant
+   * error message.
+   */
+  private @Nullable String validateNewFile(Path newJsonFile) {
+    Objects.requireNonNull(newJsonFile);
+
+    try (InputStream inputStream = Files.newInputStream(newJsonFile)) {
+      Results results = objectMapper.readValue(inputStream, Results.class);
+
+      if (results.testMetadata == null || results.testMetadata.isEmpty()) {
+        return "Results must contain non-empty test metadata";
+      }
+    } catch (JsonProcessingException e) {
+      logger.warn("Exception processing json file {}", newJsonFile, e);
+      return "Invalid results JSON";
+    } catch (IOException e) {
+      logger.warn("Exception reading json file {}", newJsonFile, e);
+      return "Error reading results.";
+    }
+
+    return null;
+  }
+
+  /**
+   * Ensure that a results file of the given size isn't too big, and
+   * that the share directory is not full.
+   *
+   * @param resultsJsonSizeBytes The size of the results file to be uploaded,
+   *                             in bytes.
+   * @throws ShareResultsUploadException If the file is too large. or the
+   * share directory is at capacity.
+   */
+  private void validateUploadSize(long resultsJsonSizeBytes)
+      throws ShareResultsUploadException {
+    if (resultsJsonSizeBytes > fileStoreConfig.maxShareFileSizeBytes) {
+      throw new ShareResultsUploadException("Share uploads cannot exceed "
+          + fileStoreConfig.maxShareFileSizeBytes + " bytes.");
+    }
+
+    // We are only checking if the share directory is currently under its max
+    // size, without the addition of the new file. This reduces the complexity
+    // and potentially wasted time of compressing the json file to a zip file
+    // before checking if it can fit in the share directory. This is a fine
+    // compromise because it means that at most the share directory will exceed
+    // the max size by just one large results json file, and we will not accept
+    // further uploads after that.
+    long shareDirectorySize = FileStore.directorySizeBytes(
+        fileStore.shareDirectory().toFile());
+    if (shareDirectorySize >= fileStoreConfig.maxShareDirectorySizeBytes) {
+      throw new ShareResultsUploadException("Share uploads has reached max capacity.");
```

Comment:
Have you thought about what we (the maintainers of tfb-status) should do in this case?

Reply:
We may want to audit the share directory and ensure it didn't get to capacity due to someone trying to break or exploit this feature. Depending, we can delete data from such a user, or expand the capacity if we feel it's being used to good end. If the oldest files are quite old, we can go the route of backing up then deleting old files from the share directory. Does that make sense? Also, should this send some sort of alert email if the capacity is hit?

Reply:
An alert email would be nice.  Could you add that functionality along with a test?  We'd need spam prevention measures, perhaps something like the one-email-per-day check implemented by `RunCompleteMailer`.

I suppose a manual solution is best with respect to not losing data.  In practice, I'm not sure how I'd tell whether the uploaded results were legitimate or not.  I might address the problem by simply deleting the share directory.

Please make the email message include at least one suggestion for what we should do to fix the problem.

Reply:
I will add the email alert and related testing.

We could look at how closely together in time the results were uploaded, and if all of the results uploaded are identical. However these would be obvious signs that a user could get around. Simply deleting the share directory would fine in either case.